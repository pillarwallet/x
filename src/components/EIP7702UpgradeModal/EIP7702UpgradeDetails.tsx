import { useRef } from 'react';

// types
import { UpgradeStatus } from './types';

// components
import EIP7702UpgradeProgressStep from './EIP7702UpgradeProgressStep';

// utils
import { formatStepTimestamp } from '../../utils/upgrade';

interface EIP7702UpgradeDetailsProps {
  onClose: () => void;
  onCloseModal?: () => void;
  status: UpgradeStatus;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  errorDetails?: string;
}

const EIP7702UpgradeDetails = ({
  onClose,
  onCloseModal,
  status,
  submittedAt,
  pendingCompletedAt,
  errorDetails,
}: EIP7702UpgradeDetailsProps) => {
  const detailsRef = useRef<HTMLDivElement>(null);

  // Function to get step status based on overall status
  const getStepStatus = (
    step: 'Submitted' | 'Upgrading' | 'Completed'
  ): UpgradeStatus => {
    if (step === 'Submitted') {
      if (status === 'submitted') return 'upgrading'; // in progress
      if (
        status === 'upgrading' ||
        status === 'completed' ||
        status === 'failed'
      ) {
        return 'completed'; // done
      }
      return 'submitted'; // inactive
    }
    if (step === 'Upgrading') {
      if (status === 'upgrading') return 'upgrading'; // in progress
      if (status === 'completed' || status === 'failed') return 'completed'; // done
      return 'submitted'; // inactive
    }
    if (step === 'Completed') {
      if (status === 'completed') return 'completed';
      if (status === 'failed') return 'failed';
      return 'submitted'; // inactive
    }
    return 'submitted';
  };

  // Function to format timestamps for completed steps
  const getStepTimestamp = (step: 'Submitted' | 'Upgrading' | 'Completed') => {
    const stepStatus = getStepStatus(step);

    // Only show timestamp if step is completed
    if (stepStatus !== 'completed') return undefined;

    // No timestamp for the final "Completed/Failed" step itself
    if (step === 'Completed') return undefined;

    // Show timestamp for completed intermediate steps
    if (step === 'Submitted' && submittedAt) {
      return formatStepTimestamp(submittedAt, 'submitted');
    }

    if (step === 'Upgrading' && pendingCompletedAt) {
      return formatStepTimestamp(pendingCompletedAt, 'upgrading');
    }

    return undefined;
  };

  return (
    <div ref={detailsRef} className="flex flex-col gap-4 h-full w-full">
      <p className="text-xl text-white font-normal">Upgrade Details</p>

      {errorDetails && (
        <div className="bg-[#FF366C]/30 border border-[#FF366C] rounded-lg p-3">
          <p className="text-white text-sm">{errorDetails}</p>
        </div>
      )}

      <div className="flex h-full w-full rounded-[10px] border border-dashed border-[#25232D] p-3">
        <div className="flex flex-col items-center justify-center w-full">
          {/* Progress Steps */}
          <div className="flex flex-col w-full">
            {/* Submitted Step */}
            <EIP7702UpgradeProgressStep
              status={getStepStatus('Submitted')}
              label="Submitted"
              lineStatus={getStepStatus('Upgrading')}
              timestamp={getStepTimestamp('Submitted')}
            />

            {/* Upgrading Step */}
            <EIP7702UpgradeProgressStep
              status={getStepStatus('Upgrading')}
              label="Upgrading"
              lineStatus={getStepStatus('Completed')}
              timestamp={getStepTimestamp('Upgrading')}
            />

            {/* Completed Step */}
            <EIP7702UpgradeProgressStep
              status={getStepStatus('Completed')}
              label={
                getStepStatus('Completed') === 'failed' ? 'Failed' : 'Completed'
              }
              isLast
              timestamp={getStepTimestamp('Completed')}
            />
          </div>
        </div>
      </div>

      {/* Done Button */}
      <div className="w-full rounded-[10px] bg-[#121116] p-[2px_2px_6px_2px]">
        <button
          className="flex items-center justify-center w-full rounded-[8px] h-[42px] p-[1px_6px_1px_6px] bg-[#8A77FF] text-white font-normal text-[14px]"
          type="button"
          onClick={() => {
            // In completed or failed state, close the whole modal
            if (status === 'completed' || status === 'failed') {
              onCloseModal?.();
            } else {
              // Otherwise, just close details view
              onClose();
            }
          }}
        >
          Done
        </button>
      </div>
    </div>
  );
};

export default EIP7702UpgradeDetails;
