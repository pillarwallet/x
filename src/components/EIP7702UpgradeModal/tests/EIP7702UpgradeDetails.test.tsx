/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// components
import EIP7702UpgradeDetails from '../EIP7702UpgradeDetails';

// Mock dependencies
vi.mock('../../utils/upgrade', () => ({
  formatStepTimestamp: vi.fn((date: Date, step: string) => {
    if (step === 'submitted') {
      return (
        <>
          <span>Jan 1, 2023</span>
          <span> â€¢ </span>
          <span>12:00</span>
        </>
      );
    }
    return <span>12:00</span>;
  }),
}));

vi.mock('../EIP7702UpgradeProgressStep', () => ({
  default: ({ status, label, timestamp }: any) => {
    // Determine the actual displayed label based on status when label is 'Completed'
    const displayLabel =
      label === 'Completed' && status === 'failed' ? 'Failed' : label;
    // Use the original label for test ID to maintain consistency
    const testIdLabel = label;

    return (
      <div data-testid={`progress-step-${testIdLabel}`}>
        <span data-testid={`step-${testIdLabel}-status`}>{status}</span>
        <span data-testid={`step-${testIdLabel}-label`}>{displayLabel}</span>
        {timestamp && (
          <div data-testid={`step-${testIdLabel}-timestamp`}>
            {displayLabel}
          </div>
        )}
      </div>
    );
  },
}));

const baseProps = {
  onClose: vi.fn(),
  onCloseModal: vi.fn(),
  status: 'upgrading' as const,
  submittedAt: new Date('2023-01-01T00:00:00Z'),
  pendingCompletedAt: new Date('2023-01-01T00:01:00Z'),
};

describe('<EIP7702UpgradeDetails />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('Rendering', () => {
    it('renders correctly and matches snapshot', () => {
      const tree = renderer
        .create(<EIP7702UpgradeDetails {...baseProps} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders with upgrading status', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="upgrading" />);

      expect(screen.getByText('Upgrade Details')).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Submitted')).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Upgrading')).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Completed')).toBeInTheDocument();
    });

    it('renders with completed status', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="completed" />);

      expect(screen.getByText('Upgrade Details')).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Submitted')).toBeInTheDocument();
    });

    it('renders with failed status', () => {
      render(
        <EIP7702UpgradeDetails
          {...baseProps}
          status="failed"
          errorDetails="Test error"
        />
      );

      expect(screen.getByText('Upgrade Details')).toBeInTheDocument();
      expect(screen.getByText('Test error')).toBeInTheDocument();
    });

    it('displays error box when errorDetails is provided', () => {
      render(
        <EIP7702UpgradeDetails
          {...baseProps}
          errorDetails="Something went wrong"
        />
      );

      expect(screen.getByText('Something went wrong')).toBeInTheDocument();
      const errorBox = screen.getByText('Something went wrong').closest('div');
      expect(errorBox).toHaveClass('bg-[#FF366C]/30');
    });
  });

  describe('Step statuses', () => {
    it('shows correct step statuses for submitted status', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="submitted" />);

      expect(screen.getByTestId('step-Submitted-status')).toHaveTextContent(
        'upgrading'
      );
      expect(screen.getByTestId('step-Upgrading-status')).toHaveTextContent(
        'submitted'
      );
      expect(screen.getByTestId('step-Completed-status')).toHaveTextContent(
        'submitted'
      );
    });

    it('shows correct step statuses for upgrading status', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="upgrading" />);

      expect(screen.getByTestId('step-Submitted-status')).toHaveTextContent(
        'completed'
      );
      expect(screen.getByTestId('step-Upgrading-status')).toHaveTextContent(
        'upgrading'
      );
      expect(screen.getByTestId('step-Completed-status')).toHaveTextContent(
        'submitted'
      );
    });

    it('shows correct step statuses for completed status', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="completed" />);

      expect(screen.getByTestId('step-Submitted-status')).toHaveTextContent(
        'completed'
      );
      expect(screen.getByTestId('step-Upgrading-status')).toHaveTextContent(
        'completed'
      );
      expect(screen.getByTestId('step-Completed-status')).toHaveTextContent(
        'completed'
      );
    });

    it('shows failed label for completed step when status is failed', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="failed" />);

      // Should render all progress steps
      expect(screen.getByTestId('progress-step-Submitted')).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Upgrading')).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Failed')).toBeInTheDocument();

      // The Failed step should have failed status
      expect(screen.getByTestId('step-Failed-status')).toHaveTextContent(
        'failed'
      );
      expect(screen.getByTestId('step-Failed-label')).toHaveTextContent(
        'Failed'
      );
    });
  });

  describe('Timestamps', () => {
    it('displays submitted timestamp when available', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="upgrading" />);

      // Should show timestamp for submitted step
      expect(
        screen.getByTestId('step-Submitted-timestamp')
      ).toBeInTheDocument();
    });

    it('displays pending completed timestamp when available', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="completed" />);

      // Should show timestamp for upgrading step
      expect(
        screen.getByTestId('step-Upgrading-timestamp')
      ).toBeInTheDocument();
    });

    it('does not display timestamp for final Completed step', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="completed" />);

      // Should NOT show timestamp for Completed step
      expect(
        screen.queryByTestId('step-Completed-timestamp')
      ).not.toBeInTheDocument();
    });
  });

  describe('User interactions', () => {
    it('calls onClose when done button is clicked for non-final status', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="upgrading" />);

      const doneButton = screen.getByText('Done');
      fireEvent.click(doneButton);

      expect(baseProps.onClose).toHaveBeenCalled();
      expect(baseProps.onCloseModal).not.toHaveBeenCalled();
    });

    it('calls onCloseModal when done button is clicked for completed status', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="completed" />);

      const doneButton = screen.getByText('Done');
      fireEvent.click(doneButton);

      expect(baseProps.onCloseModal).toHaveBeenCalled();
      expect(baseProps.onClose).not.toHaveBeenCalled();
    });

    it('calls onCloseModal when done button is clicked for failed status', () => {
      render(<EIP7702UpgradeDetails {...baseProps} status="failed" />);

      const doneButton = screen.getByText('Done');
      fireEvent.click(doneButton);

      expect(baseProps.onCloseModal).toHaveBeenCalled();
      expect(baseProps.onClose).not.toHaveBeenCalled();
    });

    it('handles missing onCloseModal gracefully', () => {
      const propsWithoutModalClose = {
        onClose: vi.fn(),
        status: 'completed' as const,
      };

      render(<EIP7702UpgradeDetails {...propsWithoutModalClose} />);

      const doneButton = screen.getByText('Done');
      fireEvent.click(doneButton);

      expect(propsWithoutModalClose.onClose).not.toHaveBeenCalled();
    });
  });

  describe('Edge cases', () => {
    it('handles missing timestamps gracefully', () => {
      const propsWithoutTimestamps = {
        onClose: vi.fn(),
        onCloseModal: vi.fn(),
        status: 'upgrading' as const,
      };

      render(<EIP7702UpgradeDetails {...propsWithoutTimestamps} />);

      expect(screen.getByText('Upgrade Details')).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Submitted')).toBeInTheDocument();
    });

    it('handles missing error details', () => {
      render(<EIP7702UpgradeDetails {...baseProps} />);

      // Should not display error box
      expect(
        screen.queryByText((content, element) => {
          return element?.className?.includes('bg-[#FF366C]/30') || false;
        })
      ).not.toBeInTheDocument();
    });
  });
});
