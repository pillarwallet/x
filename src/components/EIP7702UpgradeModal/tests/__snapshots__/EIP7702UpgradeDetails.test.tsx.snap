// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EIP7702UpgradeDetails /> > Rendering > renders correctly and matches snapshot 1`] = `
<div
  className="flex flex-col gap-4 h-full w-full"
>
  <p
    className="text-xl text-white font-normal"
  >
    Upgrade Details
  </p>
  <div
    className="flex h-full w-full rounded-[10px] border border-dashed border-[#25232D] p-3"
  >
    <div
      className="flex flex-col items-center justify-center w-full"
    >
      <div
        className="flex flex-col w-full"
      >
        <div
          data-testid="progress-step-Submitted"
        >
          <span
            data-testid="step-Submitted-status"
          >
            completed
          </span>
          <span
            data-testid="step-Submitted-label"
          >
            Submitted
          </span>
          <div
            data-testid="step-Submitted-timestamp"
          >
            Submitted
          </div>
        </div>
        <div
          data-testid="progress-step-Upgrading"
        >
          <span
            data-testid="step-Upgrading-status"
          >
            upgrading
          </span>
          <span
            data-testid="step-Upgrading-label"
          >
            Upgrading
          </span>
        </div>
        <div
          data-testid="progress-step-Completed"
        >
          <span
            data-testid="step-Completed-status"
          >
            submitted
          </span>
          <span
            data-testid="step-Completed-label"
          >
            Completed
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="w-full rounded-[10px] bg-[#121116] p-[2px_2px_6px_2px]"
  >
    <button
      className="flex items-center justify-center w-full rounded-[8px] h-[42px] p-[1px_6px_1px_6px] bg-[#8A77FF] text-white font-normal text-[14px]"
      onClick={[Function]}
      type="button"
    >
      Done
    </button>
  </div>
</div>
`;
